### PR TITLE
Aggregate errors from llvm-dwarfdump --verify

### PR DESCRIPTION
**Pull request by kevinfrei**
The amount and format of output from `llvm-dwarfdump --verify` makes it quite difficult to know if a change to a tool that produces or modifies DWARF is causing new problems, or is fixing existing problems. This diff adds a categorized summary of issues found by the DWARF verifier, on by default, at the bottom of the error output. 

The change includes a new `--error-display` option with 4 settings:

* `--error-display=quiet`:  Only display if errors occurred, but no details or summary are printed.
* `--error-display=summary`: Only display the aggregated summary of errors with no error detail.
* `--error-display=details`: Only display the detailed error messages with no summary (previous behavior)
* `--error-display=full`: Display both the detailed error messages and the aggregated summary of errors (the default)

I changed a handful of tests that were failing due to new output, adding the flag to use the old behavior for all but a couple. For those two I added the new aggregated output to the expected output of the test.

The `OutputCategoryAggregator` is a pretty simple little class that @clayborg suggested to allow code to only be run to dump detail if it's enabled, while still collating counts of the category. Knowing that the lambda passed in is only conditionally executed is pretty important (handling errors has to be done *outside* the lambda). I'm happy to move this somewhere else (and change/improve it) to be more broadly useful if folks would like.

### CODE REVIEW

**Comment by kevinfrei** [comment]
How does one assign a reviewer if you don't have write access? I'm 98% certain I'd like @dwblaikie to see this one...

File: llvm/include/llvm/DebugInfo/DWARF/DWARFVerifier.h [hunk 1]
```diff
@@ -30,6 +30,20 @@ class DWARFDebugAbbrev;
 class DataExtractor;
 struct DWARFSection;
 
+class OutputCategoryAggregator {
+private:
+  std::map<std::string, unsigned> Aggregation;
```
Comment (dwblaikie) [inline llvm/include/llvm/DebugInfo/DWARF/DWARFVerifier.h:35]: llvm::StringMap?

File: llvm/include/llvm/DebugInfo/DWARF/DWARFVerifier.h [hunk 2]
```diff
@@ -348,6 +363,9 @@ class DWARFVerifier {
   bool verifyDebugStrOffsets(
       StringRef SectionName, const DWARFSection &Section, StringRef StrData,
       void (DWARFObject::*)(function_ref<void(const DWARFSection &)>) const);
+
+  /// Emits any aggregate information collected, depending on the dump options
+  void summarize(bool Success);
```
Comment (dwblaikie) [inline llvm/include/llvm/DebugInfo/DWARF/DWARFVerifier.h:368]: If we have this aggregator that gets called back on every error anyway - would that allow us to remove the `Success` returns and tracking, and just track success in the aggregator instead?

File: llvm/test/tools/llvm-dwarfdump/X86/verify_file_encoding.yaml [hunk 3]
```diff
@@ -50,7 +50,11 @@
 # CHECK-NEXT:               DW_AT_decl_line   [DW_FORM_sdata] (3)
 # CHECK-NEXT:               DW_AT_call_file   [DW_FORM_sdata] (4)
 # CHECK-NEXT:               DW_AT_call_line   [DW_FORM_sdata] (5){{[[:space:]]}}
-
+# CHECK-NEXT: Verifying dwo Units...
+# CHECK-NEXT: error: Aggregated error category counts:
+# CHECK-NEXT: error: Error category 'Invalid encoding in DW_AT_decl_file' occurred 4 time(s).
```
Comment (dwblaikie) [inline llvm/test/tools/llvm-dwarfdump/X86/verify_file_encoding.yaml:55]: Not sure we need the word 'category' in here? and usually diagnostics are all lower case, I think? (though maybe `error: error 'Invalid encoding in DW_AT_decl_file' occurred 4 time(s).` is a bit hard to parse.. (perhaps these shouldn't be printed as errors? It is a bit weird having a whole multiline report each prefixed with `error:`, maybe?) (though it also shows these diagnostics don't follow the lowercase habit either... which is unfortunate))

File: llvm/lib/DebugInfo/DWARF/DWARFVerifier.cpp [hunk 4]
```diff
@@ -167,20 +167,24 @@ bool DWARFVerifier::verifyUnitHeader(const DWARFDataExtractor DebugInfoData,
...
-    if (!ValidVersion)
-      note() << "The 16 bit unit header version is not valid.\n";
-    if (!ValidType)
-      note() << "The unit type encoding is not valid.\n";
-    if (!ValidAbbrevOffset)
-      note() << "The offset into the .debug_abbrev section is "
-                "not valid.\n";
-    if (!ValidAddrSize)
-      note() << "The address size is unsupported.\n";
+    ErrorCategory.Report("Invalid Length in Unit Header", [&]() {
```
Comment (dwblaikie) [inline llvm/lib/DebugInfo/DWARF/DWARFVerifier.cpp:170]: Awkward introducing new category names when there's nothing in the diagnostic itself that tells you it's part of a category (unlike, say, clang's diagnostic groups which are rendered in the diagnostic by default with the `[-Wunused]` or whatever after the diagnostic message) - could that be addressed in some way to make it clearer how these categories relate to the specific messages? (like if you wanted to grep for all the messages under a category to diff them against a baseline if you found the number went up or down)

**Comment by dwblaikie** [comment]
> How does one assign a reviewer if you don't have write access? I'm 98% certain I'd like @dwblaikie to see this one...

mentioning will do just fine :)

**Review by dwblaikie — COMMENTED** [review]
General idea I'm more or less OK with (though, like, clang/compilers in general don't have this sort of feature - so I'm not sure it's especially strongly motivated for DWARF either) - though it starts to look more like compiler diagnostics, with diagnostic groups & at that point I wonder whether it should use syntax that's more similar to the way compiler diagnostics and diagnostic groups are rendered - and maybe even share some implementation (well, layering might be difficult there, llvm can't depend on clang - but maybe some of the diagnostics stuff got refactored out to be used by flang and so might be in some usable place... )

File: llvm/include/llvm/DebugInfo/DWARF/DWARFVerifier.h (hunk 2 above)
Comment (kevinfrei) [reply llvm/include/llvm/DebugInfo/DWARF/DWARFVerifier.h:368]: Yeah. I think the Success boolean was a holdover from an earlier design. I'll expose a couple size attributes on the aggregator class.

**Comment by kevinfrei** [comment]
> General idea I'm more or less OK with (though, like, clang/compilers in general don't have this sort of feature - so I'm not sure it's especially strongly motivated for DWARF either) - though it starts to look more like compiler diagnostics, with diagnostic groups & at that point I wonder whether it should use syntax that's more similar to the way compiler diagnostics and diagnostic groups are rendered - and maybe even share some implementation (well, layering might be difficult there, llvm can't depend on clang - but maybe some of the diagnostics stuff got refactored out to be used by flang and so might be in some usable place... )

Errors here are a little different than Clang/LLVM errors, IMO: There's no reasonable way to narrow down your output, and to make them useful, they contain a dramatic amount of detail that makes aggregating through simple scripts much harder. In general, the diagnostics output in the this tool also feels pretty inconsistent (I wanted to aggregate before changing output, just for cleanliness of the work itself).

I do feel like aggregation is quite useful here, and *much* easier to do properly than in Clang/LLVM. This is (mostly) a very "one-to-one" usage scenario: Something has produced DWARF data in some binary somewhere. Validate that it's okay. (The output currently still requires you to know which problems are catastrophic, which can be ignore safely, etc...)

File: llvm/lib/DebugInfo/DWARF/DWARFVerifier.cpp (hunk 4 above)
Comment (kevinfrei) [reply llvm/lib/DebugInfo/DWARF/DWARFVerifier.cpp:170]: This was the only instance where something specifically like this occurred. My interest in not changing output with the aggregation trumped the "make this output searchable" goal. In an earlier revision, I'd reported each 'note' as a separate issue. I'd rather flag this with a TODO, so as not to change output yet, and have a subsequent diff that makes the output more consistent, and (as you mentioned) easier to connect to the category.

File: llvm/test/tools/llvm-dwarfdump/X86/verify_file_encoding.yaml (hunk 3 above)
Comment (kevinfrei) [reply llvm/test/tools/llvm-dwarfdump/X86/verify_file_encoding.yaml:55]: This was where I just threw something together, because I assumed whatever I did would require revisiting :) 
```
error: <Error summary? Aggregated issues? Not sure what's best for the header>:
error: Invalid encoding in DW_AT_decl-file was reported 4 times(s).
error: ....
```
seems better (and less redundant)

File: llvm/lib/DebugInfo/DWARF/DWARFVerifier.cpp (hunk 4 above)
Comment (kevinfrei) [reply llvm/lib/DebugInfo/DWARF/DWARFVerifier.cpp:170]: I think, for now, I'll swap this back to individual categories to make it a bit easier to connect the category to the unique error.

File: llvm/include/llvm/DebugInfo/DWARF/DWARFVerifier.h (hunk 1 above)
Comment (kevinfrei) [reply llvm/include/llvm/DebugInfo/DWARF/DWARFVerifier.h:35]: I reran tests with this change and then remembered why I went with std::map. It's sorted. llvm::StringMap isn't (hash-based). Add a single new category to the report and everything (might) get shuffled around, so I'd prefer to keep it a std::map.

File: llvm/lib/DebugInfo/DWARF/DWARFVerifier.cpp [hunk 5]
```diff
@@ -240,30 +285,38 @@ unsigned DWARFVerifier::verifyUnitContents(DWARFUnit &Unit,
...
 
   uint8_t UnitType = Unit.getUnitType();
   if (!DWARFUnit::isMatchingUnitTypeAndTag(UnitType, Die.getTag())) {
-    error() << "Compilation unit type (" << dwarf::UnitTypeString(UnitType)
-            << ") and root DIE (" << dwarf::TagString(Die.getTag())
-            << ") do not match.\n";
+    ErrorCategory.Report("Mismatched unit type", [&]() {
+      error() << "Compilation unit type (" << dwarf::UnitTypeString(UnitType)
+              << ") and root DIE (" << dwarf::TagString(Die.getTag())
+              << ") do not match.\n";
```
Comment (clayborg) [inline llvm/lib/DebugInfo/DWARF/DWARFVerifier.cpp:308]: show offset in detail or dump the Die with `Die.dump(OS)`

File: llvm/lib/DebugInfo/DWARF/DWARFVerifier.cpp [hunk 6]
```diff
@@ -240,30 +285,38 @@ unsigned DWARFVerifier::verifyUnitContents(DWARFUnit &Unit,
...
   }
 
   //  According to DWARF Debugging Information Format Version 5,
   //  3.1.2 Skeleton Compilation Unit Entries:
   //  "A skeleton compilation unit has no children."
   if (Die.getTag() == dwarf::DW_TAG_skeleton_unit && Die.hasChildren()) {
-    error() << "Skeleton compilation unit has children.\n";
+    ErrorCategory.Report("Skeleton CU has children", [&]() {
+      error() << "Skeleton compilation unit has children.\n";
+    });
```
Comment (clayborg) [inline llvm/lib/DebugInfo/DWARF/DWARFVerifier.cpp:319]: show offset of skeleton die in detail or dump the Die with `Die.dump(OS)`

File: llvm/tools/llvm-dwarfdump/llvm-dwarfdump.cpp [hunk 7]
```diff
@@ -276,6 +276,14 @@ static cl::opt<bool>
...
 static opt<bool> Verify("verify", desc("Verify the DWARF debug info."),
                         cat(DwarfDumpCategory));
+static opt<bool> OnlyAggregateErrors(
+    "only-aggregate-errors",
+    desc("Only display the aggregate errors when verifying."),
+    cat(DwarfDumpCategory));
+static opt<bool> NoAggregateErrors(
+    "no-aggregate-errors",
+    desc("Do not display the aggregate errors when verifying."),
+    cat(DwarfDumpCategory));
```
[the remark this answers is not in the document]
Comment (kevinfrei) [reply llvm/tools/llvm-dwarfdump/llvm-dwarfdump.cpp:286]: I'll also add a "no-summary" option to retain the ability to keep the old output for now.

File: llvm/lib/DebugInfo/DWARF/DWARFVerifier.cpp (hunk 5 above)
Comment (kevinfrei) [reply llvm/lib/DebugInfo/DWARF/DWARFVerifier.cpp:308]: I want to keep detailed message changes separate from the summary details. I'll improve details in a subsequent change.

File: llvm/lib/DebugInfo/DWARF/DWARFVerifier.cpp [hunk 8]
```diff
@@ -240,30 +285,38 @@ unsigned DWARFVerifier::verifyUnitContents(DWARFUnit &Unit,
...
     NumUnitErrors++;
     return NumUnitErrors;
   }
 
   if (!dwarf::isUnitType(Die.getTag())) {
-    error() << "Compilation unit root DIE is not a unit DIE: "
-            << dwarf::TagString(Die.getTag()) << ".\n";
+    ErrorCategory.Report("Compilation unit root DIE is not a unit DIE", [&]() {
+      error() << "Compilation unit root DIE is not a unit DIE: "
+              << dwarf::TagString(Die.getTag()) << ".\n";
```
[the remark this answers is not in the document]
Comment (kevinfrei) [reply llvm/lib/DebugInfo/DWARF/DWARFVerifier.cpp:298]: I want to keep detailed message changes separate from the summary details. I'll improve details in a subsequent change.

File: llvm/lib/DebugInfo/DWARF/DWARFVerifier.cpp [hunk 9]
```diff
@@ -240,30 +285,38 @@ unsigned DWARFVerifier::verifyUnitContents(DWARFUnit &Unit,
 
   DWARFDie Die = Unit.getUnitDIE(/* ExtractUnitDIEOnly = */ false);
   if (!Die) {
-    error() << "Compilation unit without DIE.\n";
+    ErrorCategory.Report("Compilation unit missing DIE", [&]() {
+      error() << "Compilation unit without DIE.\n";
```
[the remark this answers is not in the document]
Comment (kevinfrei) [reply llvm/lib/DebugInfo/DWARF/DWARFVerifier.cpp:289]: I want to keep detailed message changes separate from the summary details. I'll improve details in a subsequent change.

File: llvm/lib/DebugInfo/DWARF/DWARFVerifier.cpp (hunk 6 above)
Comment (kevinfrei) [reply llvm/lib/DebugInfo/DWARF/DWARFVerifier.cpp:319]: I want to keep detailed message changes separate from the summary details. I'll improve details in a subsequent change.

**Comment by JDevlieghere** [comment]
> The amount and format of output from `llvm-dwarfutil --verify` makes it quite difficult to know if a change to a tool that produces or modifies DWARF is causing new problems, or is fixing existing problems. 

The way `dsymutil` solves this problem is by having different verification modes. 

```
  --verify-dwarf <verification mode>
                          Run the DWARF verifier on the input and/or output. Valid options are 'none, 'input', 'output', 'all' or 'auto' which runs the output verifier only if input verification passed.
```

Most of the time, invalid DWARF in means invalid DWARF out. 

> This diff adds a categorized summary of issues found by the DWARF verifier, on by default, at the bottom of the error output.

I didn't look at the patch, but from the description I'm not sure how aggregated error output solves this problem. 

Speaking from experience with `dsymutil`, I don't think you want to have the verifier on by default. It's pretty slow and trades off performance to have better error reporting. For `dsymutil`, I made the `auto` verification mode the default in `assert` builds. It's way too slow to turn on in production.

**Comment by kevinfrei** [comment]
> > The amount and format of output from `llvm-dwarfutil --verify` makes it quite difficult to know if a change to a tool that produces or modifies DWARF is causing new problems, or is fixing existing problems.
> 
> The way `dsymutil` solves this problem is by having different verification modes.
> 
> ```
>   --verify-dwarf <verification mode>
>                           Run the DWARF verifier on the input and/or output. Valid options are 'none, 'input', 'output', 'all' or 'auto' which runs the output verifier only if input verification passed.
> ```
> 
> Most of the time, invalid DWARF in means invalid DWARF out.
> 
> > This diff adds a categorized summary of issues found by the DWARF verifier, on by default, at the bottom of the error output.
> 
> I didn't look at the patch, but from the description I'm not sure how aggregated error output solves this problem.
> 
> Speaking from experience with `dsymutil`, I don't think you want to have the verifier on by default. It's pretty slow and trades off performance to have better error reporting. For `dsymutil`, I made the `auto` verification mode the default in `assert` builds. It's way too slow to turn on in production.

`dsymutil` is a DWARF processor. `dwarfdump -verify` only does validation. The problem is that DWARF data appears almost always _slightly_ bad. For example, running llvm-dwarfdump -verify *on* llvm-dwarfdump on my linux development box spits out over 160MB of log. The error information is both overwhelming, and due to the details necessary for diagnosis, nearly impossible to "diff". So, for a tool like Bolt, they can't easily know if they've done something horribly wrong with DWARF data, or if they just moved the errors around. With an error summary, they can figure out what things they might have made worse, what new errors they may have caused, and potentially also identified problems that came from the compiler.

BTW: Here's the bottom of that --verify log. 123 ranges aren't contained by parent ranges. I hope to improve the diagnostics so the output is a bit more opinionated/informative. I believe that overlapping DW_AT_ranges, for example, just aren't that big of a deal...

```
error: Aggregated error counts:
error: DIE address ranges are not contained by parent ranges occurred 123 time(s).
error: DIE has overlapping DW_AT_ranges occurred 119212 time(s).
error: DIEs have overlapping address ranges occurred 118323 time(s).
Errors detected.
```

**Comment by dwblaikie** [comment]
> > > The amount and format of output from `llvm-dwarfutil --verify` makes it quite difficult to know if a change to a tool that produces or modifies DWARF is causing new problems, or is fixing existing problems.
> > 
> > 
> > The way `dsymutil` solves this problem is by having different verification modes.
> > ```
> >   --verify-dwarf <verification mode>
> >                           Run the DWARF verifier on the input and/or output. Valid options are 'none, 'input', 'output', 'all' or 'auto' which runs the output verifier only if input verification passed.
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Most of the time, invalid DWARF in means invalid DWARF out.
> > > This diff adds a categorized summary of issues found by the DWARF verifier, on by default, at the bottom of the error output.
> > 
> > 
> > I didn't look at the patch, but from the description I'm not sure how aggregated error output solves this problem.
> > Speaking from experience with `dsymutil`, I don't think you want to have the verifier on by default. It's pretty slow and trades off performance to have better error reporting. For `dsymutil`, I made the `auto` verification mode the default in `assert` builds. It's way too slow to turn on in production.
> 
> `dsymutil` is a DWARF processor. `dwarfdump -verify` only does validation. The problem is that DWARF data appears almost always _slightly_ bad. For example, running llvm-dwarfdump -verify _on_ llvm-dwarfdump on my linux development box spits out over 160MB of log. The error information is both overwhelming, and due to the details necessary for diagnosis, nearly impossible to "diff". So, for a tool like Bolt, they can't easily know if they've done something horribly wrong with DWARF data, or if they just moved the errors around. With an error summary, they can figure out what things they might have made worse, what new errors they may have caused, and potentially also identified problems that came from the compiler.
> 
> BTW: Here's the bottom of that --verify log. 123 ranges aren't contained by parent ranges. I hope to improve the diagnostics so the output is a bit more opinionated/informative. I believe that overlapping DW_AT_ranges, for example, just aren't that big of a deal...
> 
> ```
> error: Aggregated error counts:
> error: DIE address ranges are not contained by parent ranges occurred 123 time(s).
> error: DIE has overlapping DW_AT_ranges occurred 119212 time(s).
> error: DIEs have overlapping address ranges occurred 118323 time(s).
> Errors detected.
> ```

These are mostly bugs that should be fixed in the verifier related to address tombstoning. (three CUs with a copy of an inline function - one gets picked, the other two get marked as "dead" usually by setting the starting address to 0 (depends on the linker) and then you have two CUs with functions that start at 0, so they're "overlapping" - or, if the inline functions happen to be truly identical, both CUs might point to the one copy, again, overlapping)

It might just not be the most useful tool to use for what you're trying to do without further work to fix these false positives.

**Comment by JDevlieghere** [comment]
> dsymutil is a DWARF processor. dwarfdump -verify only does validation. 

Ah, I think there's some confusion here. Your PR descriptions says `llvm-dwarfutil` but based on your reply you were talking about `llvm-dwarfdump`. The former is the equivalent of `dsymutil` for ELF and I thought it was using the DWARF verifier as a library (like `dsymutil` does) and all my comments were related to that. However you're trying to improve the output of `llvm-dwarfdump` (which is great!) so that explains why it wasn't clear to me how that would work :-)

**Comment by clayborg** [comment]
> > BTW: Here's the bottom of that --verify log. 123 ranges aren't contained by parent ranges. I hope to improve the diagnostics so the output is a bit more opinionated/informative. I believe that overlapping DW_AT_ranges, for example, just aren't that big of a deal...
> > ```
> > error: Aggregated error counts:
> > error: DIE address ranges are not contained by parent ranges occurred 123 time(s).
> > error: DIE has overlapping DW_AT_ranges occurred 119212 time(s).
> > error: DIEs have overlapping address ranges occurred 118323 time(s).
> > Errors detected.
> > ```
> 
> These are mostly bugs that should be fixed in the verifier related to address tombstoning. (three CUs with a copy of an inline function - one gets picked, the other two get marked as "dead" usually by setting the starting address to 0 (depends on the linker) and then you have two CUs with functions that start at 0, so they're "overlapping" - or, if the inline functions happen to be truly identical, both CUs might point to the one copy, again, overlapping)

We should fix the DWARFVerifier to do what llvm-gsymutil does: it gets the list of valid ".text" ranges. It does this by scanning the sections and making a set of address ranges that have read + execute permissions, and only checking addresses for any DIEs if they fall into these ranges. This keeps the GSYM file from having extra invalid addresses in the output file. Code is easy:

```
  AddressRanges TextRanges;
  for (const object::SectionRef &Sect : Obj.sections()) {
    if (!Sect.isText())
      continue;
    const uint64_t Size = Sect.getSize();
    if (Size == 0)
      continue;
    const uint64_t StartAddr = Sect.getAddress();
    TextRanges.insert(AddressRange(StartAddr, StartAddr + Size));
  }
```

> 
> It might just not be the most useful tool to use for what you're trying to do without further work to fix these false positives.

Yes, I can work with Kevin to fix these if you are ok with the above text ranges solution? I works really well for GSYM.

**Comment by kevinfrei** [comment]
> great

ZOMG, you are correct: I typed "llvm-dwarfutil" when I meant "llvm-dwarfdump". Fixed 🤡

**Comment by kevinfrei** [comment]
> Your PR descriptions says `llvm-dwarfutil` but based on your reply you were talking about `llvm-dwarfdump`

ZOMG, you are correct: I typed "llvm-dwarfutil" when I meant "llvm-dwarfdump". Fixed 🤡

**Comment by kevinfrei** [comment]
I also updated the PR description to include the changed flags that @clayborg suggested.

**Comment by kevinfrei** [comment]
Ping @JDevlieghere  or @dwblaikie for an approval and clicking ye-old-purple-button-of-squashing-and-rebasing. I'm hoping to get something similar into llvm-gsymutil next (while procrastinating migration of Debuginfod tests from shell to api in LLDB🥇)

**Review by clayborg — COMMENTED** [review]
LGTM, but I will let other LLVM folks give the final OK.

**Review by JDevlieghere — APPROVED** [review]
LGTM!

**Comment by dwblaikie** [comment]
I'm ok with @JDevlieghere's sign off here.
